text/3389-manifest-lint.md

```diff
@@ -0,0 +1,291 @@
+
+Instead of traditional workspace inheritance where there is a single value to
+inherit with `workspace = true`, we could have `[workspace.lints.<preset>]`
+which defines presets and the user could do `lints.<preset> = true`.  The user
+could then name them as they wish to avoid collision with rustc lints.
+
+Instead of the `[lints]` table being `lint = "level"`, we could organize
+it around `level = ["lint", ...]` like some other linters do (like
+[ruff](https://beta.ruff.rs/docs/configuration/)) but this works better for
+logically organizing lints, highlighting what changed in diffs, and for
+possibly adding lint-specific configuration in the future.
+
+# Prior art
+[prior-art]: #prior-art
+
+Rust
+- [cargo cranky](https://github.com/ericseppanen/cargo-cranky)
+
+Python
+- [flake8](https://flake8.pycqa.org/en/latest/user/configuration.html)
+  - Format is `level = [lint, ...]`
+- [pylint](https://github.com/PyCQA/pylint/blob/main/examples/pylintrc#L402)
+  - Format is `level = [lint, ...]` but the [config file is a reflection of the CLI](https://pylint.pycqa.org/en/latest/user_guide/configuration/index.html)
+- [ruff](https://beta.ruff.rs/docs/configuration/)
+  - Format is `level = [lint, ...]`, likely due to past precedence in ecosystem (see above)
+
+Javascript
+- [eslint](https://eslint.org/docs/latest/use/configure/rules)
+  - Format is `lint = level` / `lint = [ level, additional config ]`
+
+Go
+- [golangci-lint](https://golangci-lint.run/usage/configuration/)
+  - Format is `level = [lint, ...]`
+
+Ruby
+- [rubocop](https://docs.rubocop.org/rubocop/1.45/configuration.html)
+  - Format is `Lint: Enabled: true`
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+How should we hand rustdoc lint levels or, in the future, cargo lint levels?
+The current proposal takes all lints and passes them to rustc like `RUSTFLAGS`
+but rustdoc uses `RUSTDOCFLAGS` and cargo would use neither.  This also starts
+to get into
+[user-defined tool attributes](https://rust-lang.github.io/rfcs/2103-tool-attributes.html).
+
+Should we only apply/fingerprint lints for the appropriate tool?  For example,
+we would not include and fingerprint `clippy::` lints when running builds,
+allowing them to change without forcing a rebuild.  We likely already need to
+be tool-aware for built-in tools to handle `rustdoc::` lints (see above) so
+this isn't much more of a step.
+
+How do we allow controlling precedence between lints and lint groups?  We are
+using a TOML table with the keys as lint names which does not allow controlling
+ordering.  Even if we switched to `level = [lint, ...]`, you get a hard coded
+precedence between levels that the user can't control.
+
+What does this mean for rustfmt?  If this feature slowly absorbs the role of
+`clippy.toml`, a `rustfmt.toml` file might stick out like a sore thumb.  Do we
+accept that, shoe-horn it in, generalize this feature into "tools" and "rules"
+from "linters" and "lints"?
```

Comment:
@CobaltCause 

> > should we generalize the RFC to be about "tools" and "rules", rather than "linters" and "lints"?
> 
> I'm inclined to think so. To me, "linters" and "lints" seems like a special case of "tools" and "configuration" (not totally sure what's meant by "rules"). On this topic, I made [this comment](https://github.com/rust-lang/cargo/issues/5034#issuecomment-1244722861) on the cargo issue, pointing out [Python's `pyproject.toml` `[tool]` section from PEP 518](https://peps.python.org/pep-0518/#tool-table) as prior art for this idea. I think tools like [`cargo-audit`](https://github.com/rustsec/rustsec/tree/main/cargo-audit) and [`cargo-embed`](https://github.com/probe-rs/probe-rs/tree/master/cargo-embed) would benefit from a general solution like that (if not from `workspace-deduplicate`, at least from not having a bunch of config files laying around).

Reply:
There was a [prior proposal for a `tool` section](https://internals.rust-lang.org/t/tool-configs-in-cargo-toml-e-g-rustfmt-clippy-and-alternatives-brainstorming/9055).

In the linked issue, you mentioned

> [Cargo already has [package.metadata] and [workspace.metadata]](https://doc.rust-lang.org/cargo/reference/manifest.html#the-metadata-table) too which sound pretty similar.

For me, the question is what is the value add of generalizing `lint` to `tool` in light of us having `metadata`?

I think the value of `lint` is that it is more structured, closed-ended design where cargo can more specifically act on it.

Reply:
Perhaps there's a third option where you have `[lint]` but it is only for rustc lints. Clippy lint config can be under `[tool.clippy.lint]`. That is, each tool can define its own lint config (that should follow some conventions).

Reply:
> There was a [prior proposal for a `tool` section](https://internals.rust-lang.org/t/tool-configs-in-cargo-toml-e-g-rustfmt-clippy-and-alternatives-brainstorming/9055).

Interesting, I hadn't seen that proposal before; the author articulates their reasoning a lot better than I could have :sweat_smile:

> For me, the question is what is the value add of generalizing `lint` to `tool` in light of us having `metadata`?

Reading through [the documentation for `{package,workspace}.metadata`](https://doc.rust-lang.org/cargo/reference/manifest.html#the-metadata-table) and [the `workspace-deduplicate` RFC](https://rust-lang.github.io/rfcs/2906-cargo-workspace-deduplicate.html?highlight=2906#placeholder-values), I think the technical value-add is that `{workspace,package}.metadata` don't support workspace inheritance whereas this RFC does. There's also the social value-add of this RFC, which is that it uses stronger language (like "is" instead of "can"), which might lead to wider adoption (assuming a reason we don't see `{workspace,package}.metadata` being used broadly is due to uncertainty in tool developers about whether they should use those).

> I think the value of `lint` is that it is more structured, closed-ended design where cargo can more specifically act on it.

I guess I'm missing something because this makes it sound like some functionality would be lost if the current state of this RFC were generalized; is that the case?

Reply:
> Reading through [the documentation for {package,workspace}.metadata](https://doc.rust-lang.org/cargo/reference/manifest.html#the-metadata-table) and [the workspace-deduplicate RFC](https://rust-lang.github.io/rfcs/2906-cargo-workspace-deduplicate.html?highlight=2906#placeholder-values), I think the technical value-add is that {workspace,package}.metadata don't support workspace inheritance whereas this RFC does. T

The more open-ended we make this, the less flexibility we have.  For example, a challenge with doing workspace inheritance with `metadata` is that its design is left open ended and we might start interpreting things as they weren't intended.  Also, one of the future possibilities in this RFC is to extend workspace inheritance so the two levels can be merged.  It would be difficult to support that with an open ended design as we don't know enough about the data to correctly merge it.

> (assuming a reason we don't see {workspace,package}.metadata being used broadly is due to uncertainty in tool developers about whether they should use those).

I think we should ask them, rather than assume why they don't use it.

Also as a counter example, I do support it in `cargo release`.  I do implicit workspace inheritance though as I supported that before the workspace inheritance feature was implemented.

> I guess I'm missing something because this makes it sound like some functionality would be lost if the current state of this RFC were generalized; is that the case?

I think we could generalize the shape of the data but if we generalize it too far, then this just becomes another `metadata` table.

Reply:
I will also speak into the rustfmt specifics of this, but I first want to underscore my view that it's important to consider this holistically around the experience and expectations developers may have for configuration across the various official tools if we move ahead with this proposal.

I do not have the receipts handy, but I'm sure many have seen/heard comments to the effect of "I have `edition` set to `2018` in Cargo.toml, but when I run `rustc ...` it's using edition 2015 instead of edition 2018".

Essentially, it's a relatively common occurrence for some developers to (incorrectly) assume that info defined in Cargo.toml is honored by all tools even outside a cargo context. While this isn't necessarily a problem for tools typically, or always, invoked as part of some `cargo ...` subcommand flow, it's a persistent topic that pops up for tools that are utilized from both cargo and non-cargo contexts (`rustc`, `rustdoc`, `rustfmt`).

Fundamentally, my concern is that supporting more configuration values in Cargo.toml could exacerbate this issue. If this proceeds, then I'd love to find a way to help ameliorate this by somehow increasing awareness of when values in Cargo.toml can be used/will be honored, and crucially, when they won't. (though unfortunately I've no prescriptions).

<hr />

As for rustfmt, quick overview for anyone unfamiliar

`rustfmt` can and often is invoked in a cargo context via `cargo fmt`, but it's also commonly invoked directly both by humans and machines (e.g. editor/ide use cases to format single file, subset of a single file, etc.). In the `cargo fmt` scenario obviously relevant Cargo.toml info is utilized, but `rustfmt ...` itself doesn't, so when running rustfmt directly, one needs to explicitly specify the edition to use a non-2015 edition. rustfmt also supports various [config options](https://rust-lang.github.io/rustfmt/?version=v1.5.1&search=) that can be specified in a `rustfmt.toml` file.

We've had requests to support allowing rustfmt config options to also be defined in Cargo.toml, but it's something we've always pushed back on for a number of reasons, one of which being that `rustfmt` would either have to ignore the edition value in Cargo.toml while honoring other aspects of Cargo.toml, or essentially introduce breaking behavior where the default/implicit edition rustfmt would utilize would change in some contexts (since the ecosystem default is still 2015 and rustfmt does not currently utilize Cargo.toml defined edition, like rustc and rustdoc)

<hr />

I understand this proposal is focused on lints, and it's not my intent to scope creep. I would, however, feel more comfortable if we could either set an upper bound on how far this could extend to other tools (explicitly, even if only notionally, to not expand beyond lints), or determine if this could/should expand, and if so then try to address the potential challenges this could present for other tools

Reply:
This reminds me of how "restic" (https://github.com/restic/restic) handles this. They only support command-line options (and env variables) with the rationale that config files are something that can easily be implemented by wrappers and doesn't need support by the tool.

Initially, I was annoyed by the lack of a config file but it makes sense from a maintainer perspective. Users can build wrappers with a config file format that suits their needs.

Translating this into Rust land, it would mean:

- Low-level tools like rustc, rustfmt, rustdoc, clippy etc should not support config files
- Wrapper tools like cargo can define a config file

One challenge to solve here would be how e.g. lint-specific config values are passed on the command-line.

Reply:
> The more open-ended we make this, the less flexibility we have. For example, a challenge with doing workspace inheritance with `metadata` is that its design is left open ended and we might start interpreting things as they weren't intended. Also, one of the future possibilities in this RFC is to extend workspace inheritance so the two levels can be merged. It would be difficult to support that with an open ended design as we don't know enough about the data to correctly merge it.

Okay, this answers my question about the possibility of lost functionality when increasing generality, especially the point about merging configuration (the answer is "yes").

> I think we should ask them, rather than assume why they don't use it.

Good point...

> I do implicit workspace inheritance though as I supported that before the workspace inheritance feature was implemented.

I think this is a reasonable solution for workspace inheritance without real workspace inheritance, since like you said, if the structure is too general then there's no way for `cargo` to know how to correctly do the merging. Doing it this way lets each tool that wants this decide whether and how for themselves, which is more flexible.

---

> Fundamentally, my concern is that supporting more configuration values in Cargo.toml could exacerbate this issue.

I think this is already going to be a problem for `rustc` and `[lints.rust]` as defined in the current state of this RFC, so it would probably be good to "find a way to help ameliorate this by somehow increasing awareness of when values in Cargo.toml can be used/will be honored, and crucially, when they won't" either way. (I don't have any prescriptions here either.)

> `rustfmt` would either have to ignore the edition value in Cargo.toml while honoring other aspects of Cargo.toml, or essentially introduce breaking behavior where the default/implicit edition rustfmt would utilize would change in some contexts (since the ecosystem default is still 2015 and rustfmt does not currently utilize Cargo.toml defined edition, like rustc and rustdoc)

Hmm, this is unfortunate. I wonder if a boolean option like `respect_edition_from_cargo_dot_toml` in `rustfmt` that defaults to `false` (then document this as being recommended for projects that also use `cargo`) would help?

---

I've suddenly realized what the "'tools' and 'rules'" option was about, and I think this would be a good middle ground. My revised thinking is that this RFC should consider changing from `[lints]` to `[rules]` and keep pretty much everything else the same. I think `[rules]` is just enough additional generality over `[lints]` because "lint" usually means the effect is just that some warning message(s) will be generated at some point, whereas a "rule" may actually be acted upon in some general way, for example `rustfmt` formatting a file.

This should work with workspace inheritance because this RFC already makes space for that, and it should also work with the future possibility of "extend[ing] workspace inheritance so the two levels can be merged". I'm imagining this would happen like this:

```toml
# At the workspace level
[rules.some_tool]
rule_a = "x"
rule_b = "y"
```
\+
```toml
# At the package level
[rules.some_tool]
# Maybe with a `workspace = true` here as well
rule_b = { value = "z", option = true }
rule_c = "w"
```
=
```toml
# Resultant merged rules for the package
rule_a = "x"
rule_b = { value = "z", option = true }
rule_c = "w"
```

Where the right hand side of `=` in the TOML is treated as just a blob and updated wholesale, as shown.

I think this behavior could be useful for more things than just linters, like maybe parts of `cargo-audit` and definitely `rustfmt` (in a hypothetical world where `rustfmt` wants to support this, I guess).

Reply:
Not going to have time for all of this right now but wanted to call this out

> I've suddenly realized what the "'tools' and 'rules'" option was about, and I think this would be a good middle ground. 

I got the name "rules" from [eslint](https://eslint.org/docs/latest/use/configure/rules).  Ironic for me to not call out the prior art in the threads as I'm barely familiar with that ecosystem (it is more generically linked to in the RFC).  As that, on a surface level, seems to be both tool-agnostic while still being linter-related, I figure that is the closest prior art to what we are looking at.  I should read up further on how its designed.  If anyone has experience with it, I'd also be interested in hearing how eslint's configuration system worked out for people.

Reply:
> We've had requests to support allowing rustfmt config options to also be defined in Cargo.toml, but it's something we've always pushed back on for a number of reasons, one of which being that rustfmt would either have to ignore the edition value in Cargo.toml while honoring other aspects of Cargo.toml, or essentially introduce breaking behavior where the default/implicit edition rustfmt would utilize would change in some contexts (since the ecosystem default is still 2015 and rustfmt does not currently utilize Cargo.toml defined edition, like rustc and rustdoc)

My expectation would be that invoking `rustfmt` directly behaves the same as `rustc`/`rustdoc` and its current behavior wrt. `package.edition`, it would not read the config from `Cargo.toml` itself, `cargo fmt` would read the `Cargo.toml` and translate that into CLI options. The one inconsistency then would be the fact that `rustfmt` has a config file at all, `rustc` etc. doesn't and is only configured via CLI options and relies on wrappers like `cargo rustc` to provide the config file support.

Reply:
> cargo fmt would read the Cargo.toml and translate that into CLI options.

I recognize that on paper this sounds simple and straightforward, but in actuality it could get quite messy, quite quickly. Most (if not all) editor/ide format on save behaviors run `rustfmt ...` directly so if this were to happen I fear it would be far too easy for users to find themselves where their autosave formatting in the editor would conflict a `cargo fmt` run executed manually and/or as a CI gate, which would be maddening to say the least.

It would also complicate rustfmt's existing support for project/directory level formatting configurations since CLI config args override rustfmt.toml config.

Basically, rustfmt faced and tackled this same problem quite some time ago but opted for the opposite solution (everything in `*tool*.toml` instead of `Cargo.toml`). If the Cargo.toml route is indeed selected for clippy and rustc lint configs then I just wanted to make sure that we're either explicit about this not inherently being expected to cascade out to other tools, or if we do want a more consistent experience across tools, then at least acknowledge that some of these thorny points would need to be addressed before that could happen

Reply:
I think we should keep this RFC about a `[lints]` section and clearly separate it from a tools configuration section. IMO this is a whole different topic ...

> One challenge to solve here would be how e.g. lint-specific config values are passed on the command-line.

... The main reason for this is that there _aren't_ "lint-specific config values", at least not in Clippy. Some of the config values you can set in `clippy.toml` might only affect a single lint, but it still configures the tool, that then changes its behavior (for one lint), rather than the lint directly.

More details: https://github.com/rust-lang/rfcs/pull/3389#discussion_r1108211518

> Basically, rustfmt faced and tackled this same problem quite some time ago but opted for the opposite solution (everything in `*tool*.toml` instead of `Cargo.toml`)

Clippy did the same thing and has a `clippy.toml` for configuring the tool (and because the tool is a linter, that file configures lints).

---

The reason I think a `[lints]` section makes sense separated from a `[rules.<tool>]` section (or similar) is, that lints are a first class compiler feature. Tools are more a `cargo` feature[^1].

[^1]: And probably the tools will only pick up those configurations when specified in the `Cargo.toml` when run through their cargo subcommand. At least this would be the case for Clippy and IIUC also for `rustfmt`.

Then the extension of the `[lints]` section with a `[lints.<tool>]` section is analogue to tool lints, which are also a first class compiler feature.

---

So I would not classify tool configuration as a future possibility/extension of this PR[^2], but as a complete different discussion. We might or might not want to include a `metadata` section for tools in the future, but this is IMO out of the scope of this RFC. This RFC should focus solely on if and how lint **levels** can be set in the `Cargo.toml` - for builtin compiler lints and for tool lints.

[^2]: I know that in earlier threads I stated something different, but reading through the discussions and thinking about it more I changed my mind.

Reply:
> My revised thinking is that this RFC should consider changing from [lints] to [rules] and keep everything else the same. I think [rules] is just enough additional generality over [lints] because "lint" usually means the effect is just that some warning message(s) will be generated at some point, whereas a "rule" may actually be acted upon in some general way, for example rustfmt formatting a file.

This looks like a strong argument to keep them separate. They have very different behaviour and use cases for the end user. For example, an IDE would want to read the `[lints]` section and extract projectwide lint configuration for its inline warnings. If lints were just some `[rules.clippy.<lint>]` values, this could cause confusion between lints and arbitrary clippy options and would require some complex disambiguation on the IDE side. As a user, it's also very desirable to be able to tell at a glance whether the project changes lint settings in `Cargo.toml` and in what way. Finally, since lints don't change the compiler output, they could be ignored for the purposes of artifact caching. For general tooling rules this assumption may be false.

Reply:
Sorry for not coming back to this for so long, but at this point I believe that the points are much stronger that this RFC should stick with `[lints]` as-is and not try to generalize any further. I appreciate the consideration and thought put into the discussion here!